api/src/main/java/org/openmrs/module/sync/SyncUtil.java

```diff
@@ -289,212 +302,202 @@ public static Object newObject(String className) throws Exception {
-					
-					// get the type of class defined in the text node
-					// if it is different, we could be dealing with something like Cohort.memberIds
-					// node type comes through as java.util.Set<classname>
-					String nodeDefinedCollectionType = null;
-					int indexOfLT = nodeDefinedClassName.indexOf("<");
-					if (indexOfLT > 0)
-						nodeDefinedCollectionType = nodeDefinedClassName.substring(indexOfLT + 1,
-						    nodeDefinedClassName.length() - 1);
-					
-					// change the string to just a comma delimited list
-					fieldVal = fieldVal.replaceFirst("\\[", "").replaceFirst("\\]", "");
-					
-					for (String eachFieldVal : fieldVal.split(",")) {
-						eachFieldVal = eachFieldVal.trim(); // take out whitespace
-						if(!StringUtils.hasText(eachFieldVal))
-							continue;
-						// try to convert to a simple object
-						Object tmpObject = convertStringToObject(eachFieldVal, (Class) collectionType);
-						
-						// convert to an openmrs object
-						if (tmpObject == null && nodeDefinedCollectionType != null)
-							tmpObject = getOpenmrsObj(nodeDefinedCollectionType, eachFieldVal).getId();
-						
-						if (tmpObject == null)
-							log.error("Unable to convert: " + eachFieldVal + " to a " + collectionTypeClassName);
-						else
-							tmpCollection.add(tmpObject);
+					if (tmpObject == null) {
+						log.error("Unable to convert: " + eachFieldVal + " to a " + collectionTypeClassName);
 					}
-					
-					o = tmpCollection;
-				} else if (Map.class.isAssignableFrom(classType) || Properties.class.isAssignableFrom(classType)) {
-					Object tmpMap = SyncUtil.getNormalizer(classType).fromString(classType, fieldVal);
-					
-					//if we were able to convert and got anything at all back, assign it
-					if (tmpMap != null) {
-						o = tmpMap;
+					else {
+						tmpCollection.add(tmpObject);
 					}
-				} else if ((o = convertStringToObject(fieldVal, classType)) != null) {
-					log.trace("Converted " + fieldVal + " into " + classType.getName());
-				} else {
-					log.debug("Don't know how to deserialize class: " + className);
 				}
 			}
+
+			o = tmpCollection;
 		}
-		
-		if (o == null)
-			log.debug("Never found a property named: " + fieldName + " for this class");
-		
+
+		// Handle Maps and Properties
+		else if (Map.class.isAssignableFrom(classType) || Properties.class.isAssignableFrom(classType)) {
+			Object tmpMap = SyncUtil.getNormalizer(classType).fromString(classType, xmlFieldValue);
+					
+			//if we were able to convert and got anything at all back, assign it
+			if (tmpMap != null) {
+				o = tmpMap;
+			}
+		}
+
+		// If none of the special cases handled above apply, convert the string value to an object of the field type
+		else {
+			o = convertStringToObject(xmlFieldValue, classType);
+		}
+
+		if (o == null) {
+			throw new IllegalArgumentException("Unable to convert " + xmlFieldValue + " into " + classType.getName());

```

Comment:
This is another area I'm curious about.  In the previous code, if the object was null here, the code just logged a debug message (see above) that no property value was able to be found.  I'm made things more strict, such that if this class can't identify a non-null value, then we fail hard with an exception.  This caused several unit tests to fail which had been getting away with using Dummy data previously, but fixing those to use property data was easy, and everything still passes.  I think throwing the Exception here is the right behavior.  Otherwise, sync could result in properties not getting set as intended, without anyone being aware of it.